### PR TITLE
Fix SEGV when using dynamic addresses.

### LINKED ISSUE #197: Using dynamic address causes SEGV
Test code
```
func TestSendReceive(t *testing.T) {
	c, err := amqp.Dial("amqp://")
	require.NoError(t, err)
	defer c.Close()
	ssn, err := c.NewSession()
	require.NoError(t, err)
	r, err := ssn.NewReceiver(amqp.LinkAddressDynamic())
	require.NoError(t, err)
	s, err := ssn.NewSender(amqp.LinkAddress(r.Address()))
	require.NoError(t, s.Send(context.Background(), amqp.NewMessage([]byte("hello"))))
	m, err := r.Receive(context.Background())
	require.NoError(t, err)
	assert.Equal(t, "hello", string(m.GetData()))
}
```

Result:
```
--- FAIL: TestSendReceive (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x6b71dd]

goroutine 248 [running]:
testing.tRunner.func1(0xc0002cc000)
	/usr/lib/golang/src/testing/testing.go:874 +0x3a3
panic(0x7dfca0, 0xb64460)
	/usr/lib/golang/src/runtime/panic.go:679 +0x1b2
pack.ag/amqp.attachLink(0xc000290180, 0xc0002a8040, 0xc0000edf58, 0x1, 0x1, 0x0, 0x0, 0x0)
	/home/aconway/go/src/pack.ag/amqp/client.go:981 +0x54d
pack.ag/amqp.(*Session).NewReceiver(0xc000290180, 0xc0000edf58, 0x1, 0x1, 0x0, 0x0, 0x0)
	/home/aconway/go/src/pack.ag/amqp/client.go:311 +0x81
pack.ag/amqp_test.TestSendReceive(0xc0002cc000)
	/home/aconway/go/src/pack.ag/amqp/local_test.go:48 +0x1a0
testing.tRunner(0xc0002cc000, 0x86d640)
	/usr/lib/golang/src/testing/testing.go:909 +0xc9
created by testing.(*T).Run
	/usr/lib/golang/src/testing/testing.go:960 +0x350
exit status 2
FAIL	pack.ag/amqp	0.032s
```

### PR DESCRIPTION
**Pull request by alanconway**
Fixes  #197

Signed-off-by: Alan Conway <aconway@redhat.com>

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -857,6 +857,20 @@ type link struct {
...
+	}
+	return l.source
+}
+
+func (l *link) getTarget() *target {
+	if l.target == nil {
+		l.target = new(target)
+	}
+	return l.target
+}
```
Comment (vcabbage) [inline client.go]: I'm ok with init methods, but I'd like to avoid getters returning pointers that are then modified.

File: local_test.go [hunk 2]
```diff
@@ -3,20 +3,22 @@
 package amqp_test
 
 import (
+	"context"
 	"net"
 	"testing"
 
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
```
Comment (vcabbage) [inline local_test.go]: Please avoid assertion libraries.

File: local_test.go [hunk 3]
```diff
@@ -35,3 +37,28 @@ func TestDial_IPV6(t *testing.T) {
...
+	ssn, err := c.NewSession()
+	require.NoError(t, err)
+
+	r, err := ssn.NewReceiver(amqp.LinkAddressDynamic())
+	require.NoError(t, err)
+	var m *amqp.Message
+	done := make(chan error)
+	go func() {
+		var err error
+		defer func() { done <- err; close(done) }()
```
Comment (vcabbage) [inline local_test.go]: `close(done)` is unnecessary

File: local_test.go [hunk 4]
```diff
@@ -35,3 +37,28 @@ func TestDial_IPV6(t *testing.T) {
...
+		m, err = r.Receive(context.Background())
+		m.Accept()
```
Comment (vcabbage) [inline local_test.go]: This shouldn't be called if `err == nil`

File: local_test.go [hunk 5]
```diff
@@ -35,3 +37,28 @@ func TestDial_IPV6(t *testing.T) {
...
+	}()
+
+	s, err := ssn.NewSender(amqp.LinkAddress(r.Address()))
+	require.NoError(t, s.Send(context.Background(), amqp.NewMessage([]byte("hello"))))
+	require.NoError(t, <-done)
+	assert.Equal(t, "hello", string(m.GetData()))
```
Comment (vcabbage) [inline local_test.go]: I think this logic could be simplified if sending were done in the separate goroutine.

File: local_test.go [hunk 6]
```diff
@@ -35,3 +37,28 @@ func TestDial_IPV6(t *testing.T) {
...
+	var m *amqp.Message
+	done := make(chan error)
+	go func() {
+		var err error
+		defer func() { done <- err; close(done) }()
+		m, err = r.Receive(context.Background())
+		m.Accept()
+	}()
+
+	s, err := ssn.NewSender(amqp.LinkAddress(r.Address()))
```
Comment (vcabbage) [inline local_test.go]: `err` isn't checked

**Review by vcabbage — CHANGES_REQUESTED** [review]
Good catch, thank you!

File: client.go (hunk 1 above)
Comment (alanconway) [reply client.go:885]: These aren't initers, this is constructing a blank object demand. There are many places where you want to set a target/source field, but first have to make sure the target/source exists. I think a method for that is better than repeating the if target == nil {} logic everywhere - that's how the SEGV was caused - in one place someone forgot it.

File: local_test.go (hunk 2 above)
Comment (alanconway) [reply local_test.go:11]: OK.

File: local_test.go (hunk 4 above)
Comment (alanconway) [reply local_test.go:57]: Quite right

File: local_test.go (hunk 3 above)
Comment (alanconway) [reply local_test.go:55]: yup

File: client.go (hunk 1 above)
Comment (vcabbage) [reply client.go:885]: As I stated, I disagree and won't be merging this.

File: local_test.go (hunk 5 above)
Comment (alanconway) [reply local_test.go:63]: You're right!

**Comment by alanconway** [comment]
Fixed all comments except for the on-demand construction getters for target/source objects, which I think are better than manually repeating the `if c.x == nil { c.x = new(X) }; c.x.foo = bar` pattern.

**Pull request closed without merging**